packages/rfw/README.md

```diff
@@ -208,7 +328,542 @@ class _ExampleState extends State<Example> {
+
+#### `ArgumentDecoders`
+
+It is common to need to decode types that are more structured than
+merely `int`, `double`, `bool`, or `String` scalars, for example,
+enums, `Color`s, or `Paint`s.
+
+The
+[`ArgumentDecoders`](https://pub.dev/documentation/rfw/latest/rfw/ArgumentDecoders-class.html)
+namespace offers some utility functions to make the decoding of such
+values consistent.
+
+For example, the `Row` definition above has some cases of enums. To
+decode them, it uses the
+[`ArgumentDecoders.enumValue`](https://pub.dev/documentation/rfw/latest/rfw/ArgumentDecoders/enumValue.html)
+method.
+
+#### Handlers
+
+The last kind of argument that widgets can have is callbacks.
+
+Since remote widget libraries are declarative and not code, they
+cannot represent executable closures. Instead, they are represented as
+events. For example, here is how the "7" button from the
+[calculator example](https://github.com/flutter/packages/blob/main/packages/rfw/example/wasm/logic/calculator.rfwtxt)
+is represented:
+
+<?code-excerpt "wasm/logic/calculator.rfwtxt (button7)"?>
+```rfwtxt
+CalculatorButton(label: "7", onPressed: event "digit" { arguments: [7] }),
+```
+
+This creates a `CalculatorButton` widget with two arguments, `label`,
+a string, and `onPressed`, an event, whose name is "digit" and whose
+arguments are a map with one key, "arguments", whose value is a list
+with one value 7.
+
+In that example, `CalculatorButton` is itself a remote widget that is
+defined in terms of a `Button`, and the `onPressed` argument
+is passed to the `onPressed` of the `Button`, like this:
+
+<?code-excerpt "wasm/logic/calculator.rfwtxt (CalculatorButton)"?>
+```rfwtxt
+widget CalculatorButton = Padding(
+  padding: [8.0],
+  child: SizedBox(
+    width: 100.0,
+    height: 100.0,
+    child: Button(
+      child: FittedBox(child: Text(text: args.label)),
+      onPressed: args.onPressed,
+    ),
+  ),
+);
+```
+
+Subsequently, `Button` is defined in terms of a `GestureDetector`, and
+the `args.onPressed` is passed to the `onTap` argument of that
+`GestureDetector` widget.
+
+When all is said and done, and the button is pressed, an event with
+the name "digit" and the given arguments is reported to the
+`RemoteWidget`'s `onEvent` callback. That callback takes two
+arguments, the event name and the event arguments.
+
+On the implementation side, in local widget libraries, arguments like
+the `onTap` of the `GestureDetector` must be turned into a Dart
+callback.
+
+The simplest kind of callback is a `VoidCallback` (no arguments, no
+return value). To turn an `event` value in a widget's arguments in the
+remote widget library into a `VoidCallback` in Dart that reports the
+event as described above, the `DataSource.voidHandler` method can be
+used. For example, here is a simplified `GestureDetector` that just
+implements `onTap`:
+
+<?code-excerpt "../test/readme_test.dart (onTap)"?>
+```dart
+    return GestureDetector(
+      onTap: source.voidHandler(['onTap']),
```

Comment:
Similarly, in these lines the name "GestureDetector" appears twice but means two different things: a local widget, and the Flutter widget.

Looking at the full example, I think this would actually gain a lot of clarity just by including a couple more lines in the excerpt:
```dart
          'GestureDetector': (BuildContext context, DataSource source) {
            return GestureDetector(
              onTap: source.voidHandler(<Object>['onTap']),
              child: source.optionalChild(<Object>['child']),
            );
          },
```

Also helpful might be more words unpacking "a simplified `GestureDetector`", like "a simplified `GestureDetector` local widget". (When trying to make sense of the text with "`GestureDetector`" meaning the Flutter widget, I tried to read that phrase as referring to the `GestureDetector` value constructed by this code's constructor call.) Or even say it's a simplification of the actual implementation of the `GestureDetector` core local widget — which I see it is — perhaps inserting an ellipsis comment like `// ... other arguments ...` in the middle.

Reply:
tried to add more context as suggested here